jax/_src/dispatch.py

```diff
@@ -283,8 +283,25 @@ def lower_xla_callable(fun: lu.WrappedFun, device, backend, name,
     in_type = tuple(unsafe_zip(abstract_args, itertools.repeat(True)))
     fun = lu.annotate(fun, in_type)
   else:
-    assert all(map(core.typematch, abstract_args,
-                   [a for a, k in fun.in_type if k]))
+    # Check that the provided abstract_args are consistent with in_type by first
```

Comment:
This check was just buggy before; we need to do some substitution to make sure the arguments to which the jaxpr is being applied (`abstract_args`) is consistent with `fun.in_type`.

Reply:
Actually, the new code here wasn't quite right either, because we don't want `abstract_args` in the cache key! This must have bitrotted while JAX+IREE was broken.